crates/common/logged_command/src/logged_command.rs

```diff
@@ -1,20 +1,48 @@
 use log::error;
+use nix::unistd::Pid;
 use std::ffi::OsStr;
+use std::io::ErrorKind;
 use std::process::Output;
 use std::process::Stdio;
+use std::time::Duration;
 use tokio::fs::File;
 use tokio::io::AsyncWriteExt;
 use tokio::io::BufWriter;
 use tokio::process::Child;
 use tokio::process::Command;
 
+#[derive(Debug)]
+pub enum CmdStatus {
+    Successful,
+    KilledWithSigterm,
+    KilledWithSigKill,
+}
 #[derive(Debug)]
 pub struct LoggingChild {
     command_line: String,
     pub inner_child: Child,
 }
 
 impl LoggingChild {
+    pub async fn wait_for_output_with_timeout(
+        self,
+        logger: &mut BufWriter<File>,
+        graceful_timeout: Duration,
+        forceful_timeout: Duration,
+    ) -> Result<Output, std::io::Error> {
+        let cid = self.inner_child.id();
+        let cmd_line = self.command_line;
+        let mut status = CmdStatus::Successful;
+        tokio::select! {
+            outcome = self.inner_child.wait_with_output() => {
+               Self::update_and_log_outcome(cmd_line, outcome, logger, graceful_timeout, &mut status).await
+            }
+            _ = Self::timeout_operation(&mut status, cid, graceful_timeout, forceful_timeout) => {
+                Err(std::io::Error::new(std::io::ErrorKind::Other,"failed to read the output"))
```

Comment:
One failed to kill the process not to read its output.

```suggestion
                Err(std::io::Error::new(std::io::ErrorKind::Other,"failed to kill the process: {cmd_line}"))
```